netlogo-gui/src/main/ide/JumpToDeclaration.scala

```diff
@@ -0,0 +1,88 @@
+    iterator.find(p => p.start < position && p.end >= position)
+  }
+
+  def jumpToDeclaration(cursorPosition: Int, editorArea: EditorArea): Unit ={
+    val tokenOption = findTokenContainingPosition(editorArea.getText(), cursorPosition)
+    for(token <- tokenOption) {
+      val t = getDeclaration(token, editorArea.getText)
+      editorArea.select(t.start, t.end)
+    }
+  }
+
+  def getDeclaration(token: Token, source: String): Token = {
+    val iterator = Femto.scalaSingleton[TokenizerInterface]("org.nlogo.lex.Tokenizer").tokenizeString(source).toSeq
+    val (globals, owns) = parseOwns(source)
+    val indexOfToken = iterator.indexWhere(t => t.start == token.start)
+    var stack = 0
+    for(i <- indexOfToken to 0 by -1) {
+      val tok = iterator(i)
+      if (tok.tpe.equals(TokenType.CloseBracket)) {
+        stack -= 1
+      }
+      if (tok.tpe.equals(TokenType.OpenBracket)) {
+        stack += 1
+      }
+      if(stack == 0 && tok.text.equalsIgnoreCase(token.text) && iterator(i - 1).text.equalsIgnoreCase("let")) {
+        return tok
+      }
+      if(tok.text.equalsIgnoreCase("to") || tok.text.equalsIgnoreCase("to-report")) {
+        // i + 1 is the function name
+        if(iterator(i + 2).tpe.equals(TokenType.OpenBracket)){
+          // Arguments found
+          var argIndex = i + 3
+          var arg = iterator(argIndex)
+          while(!iterator(argIndex).tpe.equals(TokenType.CloseBracket)){
+            if(iterator(argIndex).text.equalsIgnoreCase(token.text)) {
+              return iterator(argIndex)
+            }
+            argIndex += 1
+          }
+        }
+        globals.get(token.text.toLowerCase).foreach(t => return t)
+        owns.get(token.text.toLowerCase).foreach(t => return t)
+      }
+    }
+    globals.get(token.text.toLowerCase).foreach(t => return t)
+    owns.get(token.text.toLowerCase).foreach(t => return t)
+    null
+  }
+
+  def parseOwns(source: String) = {
+    val globals = scala.collection.mutable.Map[String, Token]()
+    val owns = scala.collection.mutable.Map[String, Token]()
+    val iterator = Femto.scalaSingleton[TokenizerInterface]("org.nlogo.lex.Tokenizer").tokenizeString(source)
+    while(iterator.hasNext) {
+      var token = iterator.next()
+      if (token.text.equalsIgnoreCase("globals")) {
+        token = iterator.next()
+        while (!token.tpe.equals(TokenType.CloseBracket)) {
+          if (token.tpe.equals(TokenType.Ident))
+            globals += token.text -> token
+          token = iterator.next()
+        }
+      }
+      if (token.text.toLowerCase.endsWith("-own")) {
+        token = iterator.next()
+        while (!token.tpe.equals(TokenType.CloseBracket)) {
+          if (token.tpe.equals(TokenType.Ident) && owns.get(token.text).isEmpty) {
+            owns += token.text -> token
+          }
+          token = iterator.next()
+        }
+      }
+    }
```

Comment:
This code looks like it would be quite efficient and it will work on basically all compiling NetLogo code. However, it isn't hard to devise some non-compiling NetLogo code which will cause it to fail with an exception: `NoSuchElementError`. This needs to either drop the `iterator`/`iterator.next()` calls everywhere and go with a some sort of list comprehension OR check `iterator.hasNext` before *every* call to `iterator.next()`.